### PR TITLE
feat: image minimizer

### PR DESCRIPTION
**Pull request by Asuka109**
# PR Details

## Description

- Refactor image compress plugin from loader to minimizer.
- Support compressing svg images by svgo.
- Remove `@svgr/webpack` from svg plugin in some cases.

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] I have updated changeset
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: packages/builder/plugin-image-compress/src/index.ts [hunk 1]
```diff
@@ -1,33 +1,27 @@
...
 
 /** Options enable by default: {@link DEFAULT_OPTIONS} */
 export const PluginImageCompress = (...options: Options[]): BuilderPlugin => ({
   name: 'builder-plugin-image-compress',
   setup(api) {
     const optsWithDefault = options.length ? options : DEFAULT_OPTIONS;
     const opts = optsWithDefault.map(opt => withDefaultOptions(opt));
-    api.modifyWebpackChain((chain, { CHAIN_ID, env }) => {
+    api.modifyWebpackChain((chain, { env }) => {
       if (env !== 'production') {
```
Comment (chenjiahan) [inline packages/builder/plugin-image-compress/src/index.ts:16]: 这里还有一个 output.disableMinimize 的情况

Comment (Asuka109) [reply packages/builder/plugin-image-compress/src/index.ts:16]: disableMinimize 的时候直接退出么

File: packages/builder/plugin-image-compress/src/index.ts [hunk 2]
```diff
@@ -1,33 +1,27 @@
...
-          .test(opt.test)
-          .use(`${CHAIN_ID.USE.IMAGE_COMPRESS}-${opt.use}`)
-          .loader(path.resolve(__dirname, './loader'))
-          .options(opt);
-      });
+      chain.optimization.minimize(true);
+      for (const opt of opts) {
+        chain.optimization
+          .minimizer(`image-compress-${opt.use}`)
+          .use(ModernJsImageMinimizerPlugin, [opt]);
```
Comment (chenjiahan) [inline packages/builder/plugin-image-compress/src/index.ts:23]: Plugin 有必要注册多次嘛，是不是单个插件也能实现

File: tests/e2e/webpack-builder/image-compress/package.json [hunk 3]
```diff
@@ -6,13 +6,16 @@
...
     "@modern-js/builder-webpack-provider": "workspace:*",
-    "@modern-js/builder-plugin-image-compress": "workspace:*",
+    "@modern-js/e2e": "workspace:*",
     "@types/node": "^14",
     "@types/react": "^18",
     "@types/react-dom": "^18",
     "typescript": "^4"
+  },
+  "dependencies": {
+    "react": "16"
```
Comment (chenjiahan) [inline tests/e2e/webpack-builder/image-compress/package.json:19]: 为啥要用 react 16

File: packages/builder/plugin-image-compress/package.json [hunk 4]
```diff
@@ -54,6 +54,7 @@
     "@modern-js/e2e": "workspace:*",
     "@scripts/vitest-config": "workspace:*",
     "@types/node": "^14",
+    "svgo": "^3.0.2",
```
Comment (chenjiahan) [inline packages/builder/plugin-image-compress/package.json:57]: devDependencies -> dependencies

File: packages/builder/plugin-image-compress/src/index.ts (hunk 1 above)
Comment (chenjiahan) [reply packages/builder/plugin-image-compress/src/index.ts:16]: 对滴

File: packages/builder/plugin-image-compress/src/index.ts (hunk 2 above)
Comment (Asuka109) [reply packages/builder/plugin-image-compress/src/index.ts:23]: 插件目前数组形式的配置天然可以对应 minimizer，而且这样的话可以更好复用 webpack 提供给 minimizer 的能力，比如标准的 matchObject 配置：

https://github.com/modern-js-dev/modern.js/blob/cb1f416643cbdc2995b8f537515c12f95a3a027f/packages/builder/plugin-image-compress/src/minimizer.ts#L53-L57

```js
ModuleFilenameHelpers.matchObject = (obj, str) => {
	if (obj.test) {
		if (!ModuleFilenameHelpers.matchPart(str, obj.test)) {
			return false;
		}
	}
	if (obj.include) {
		if (!ModuleFilenameHelpers.matchPart(str, obj.include)) {
			return false;
		}
	}
	if (obj.exclude) {
		if (ModuleFilenameHelpers.matchPart(str, obj.exclude)) {
			return false;
		}
	}
	return true;
};
```

File: packages/builder/plugin-image-compress/src/index.ts (hunk 1 above)
Comment (Asuka109) [reply packages/builder/plugin-image-compress/src/index.ts:16]: okk

File: tests/e2e/webpack-builder/image-compress/package.json (hunk 3 above)
Comment (Asuka109) [reply tests/e2e/webpack-builder/image-compress/package.json:19]: 当时写测试的时候随手填的，这个会有影响么🤔

Comment (chenjiahan) [reply tests/e2e/webpack-builder/image-compress/package.json:19]: 没影响，就是有点困惑。。

File: packages/builder/plugin-image-compress/src/index.ts (hunk 2 above)
Comment (chenjiahan) [reply packages/builder/plugin-image-compress/src/index.ts:23]: okk，那确实分开也合适的

**Review by chenjiahan — APPROVED** [review]

**Pull request merged**